packages/adminrouter/extra/src/test-harness/tests/test_service.py

```diff
@@ -553,3 +554,198 @@ def test_if_scheme_is_honoured_in_mesos_scheduler_entry(
+        mocker.send_command(
+            endpoint_id='http://127.0.0.1:8080',
+            func_name='always_bork',
+            aux_data=True)
+
+        # Make period cache refreshes so rare that they do not get into
+        # picture:
+        ar = nginx_class(cache_first_poll_delay=1200,
+                         cache_poll_period=1200,
+                         cache_expiration=1200,
+                         cache_max_age_soft_limit=1200,
+                         cache_max_age_hard_limit=1800,
+                         )
+        url = ar.make_url_from_path("/service/scheduler-alwaysthere/")
+
+        with GuardedSubprocess(ar):
+            resp = requests.get(
+                url,
+                allow_redirects=False,
+                headers=valid_user_header)
+
+        assert resp.status_code == 503
+        assert '503 Service Unavailable: invalid Marathon svcapps cache' in resp.text
+
+    def test_if_mesos_borkerage_prevents_resolving_via_mesosdns(
+            self, mocker, nginx_class, valid_user_header):
+        # Bork Mesos Mock, Make Marathon mock respond with no apps, so that AR
+        # tries to resolve via Mesos /state-summary
+        mocker.send_command(
+            endpoint_id='http://127.0.0.1:8080',
+            func_name='set_apps_response',
+            aux_data={"apps": []})
+        mocker.send_command(
+            endpoint_id='http://127.0.0.2:5050',
+            func_name='always_bork',
+            aux_data=True)
+
+        # Make period cache refreshes so rare that they do not get into
+        # picture:
+        ar = nginx_class(cache_first_poll_delay=1200,
+                         cache_poll_period=1200,
+                         cache_expiration=1200,
+                         cache_max_age_soft_limit=1200,
+                         cache_max_age_hard_limit=1800,
+                         )
+        url = ar.make_url_from_path("/service/scheduler-alwaysthere/")
+
+        with GuardedSubprocess(ar):
+            resp = requests.get(
+                url,
+                allow_redirects=False,
+                headers=valid_user_header)
+
+        assert resp.status_code == 503
+        assert '503 Service Unavailable: invalid Mesos state cache' in resp.text
+
+    def test_if_mesos_borkerage_does_not_prevent_resolving_via_marathon(
+            self, mocker, nginx_class, valid_user_header):
+        # Bork Mesos Mock, Make MesosDNS mock respond with no apps, so that AR
+        # is able to resolve only via Marathon/we are certain that it resolved
+        # via Marathon.
+        mocker.send_command(
+            endpoint_id='http://127.0.0.2:5050',
+            func_name='always_bork',
+            aux_data=True)
+        mocker.send_command(
+            endpoint_id='http://127.0.0.1:8123',
+            func_name='set_srv_response',
+            aux_data=EMPTY_SRV)
+
+        # Make period cache refreshes so rare that they do not get into
```

Comment:
> What could a cache refresh do here?
We want to be sure that the cache update happened only because of the request. This makes test more robust and repeatable. 

> Is it just one of these variables necessary to delay the first cache refresh?
All of them. I have described the inner workings of these variables in README.md.

> Perhaps move this out to a helper or fixture which is ar_rare_refresh or similar - and add a detailed description there.

It makes no sense here as we need raw AR instance, with a clear cache. I also try to avoid having dozens of fixtures that are used rarely - people will forget about them or misuse them.  I think it is better to have a generic one which covers all the use cases.